src/int_allocator.rs

```diff
@@ -29,62 +37,66 @@ impl IntAllocator {
     pub fn new() -> Self {
         let mut r = IntAllocator {
             u8_vec: Vec::new(),
-            node_vec: Vec::new(),
+            pair_vec: Vec::new(),
+            atom_vec: Vec::new(),
         };
         r.u8_vec.reserve(1024 * 1024);
+        r.atom_vec.reserve(256);
+        r.pair_vec.reserve(256);
         r.u8_vec.push(1_u8);
         // Preallocated empty list
-        r.node_vec
-            .push(NodePtr::Atom(IntAtomBuf { start: 0, end: 0 }));
+        r.atom_vec.push(IntAtomBuf { start: 0, end: 0 });
         // Preallocated 1
-        r.node_vec
-            .push(NodePtr::Atom(IntAtomBuf { start: 0, end: 1 }));
+        r.atom_vec.push(IntAtomBuf { start: 0, end: 1 });
         r
     }
 }
 
 impl Allocator for IntAllocator {
-    type Ptr = u32;
+    type Ptr = i32;
     type AtomBuf = IntAtomBuf;
 
     fn new_atom(&mut self, v: &[u8]) -> Self::Ptr {
         let start = self.u8_vec.len() as u32;
         self.u8_vec.extend_from_slice(v);
         let end = self.u8_vec.len() as u32;
-        let r = self.node_vec.len() as u32;
-        self.node_vec.push(NodePtr::Atom(IntAtomBuf { start, end }));
-        r
+        self.atom_vec.push(IntAtomBuf { start, end });
+        -(self.atom_vec.len() as i32)
     }
 
     fn new_pair(&mut self, first: Self::Ptr, rest: Self::Ptr) -> Self::Ptr {
-        let r: u32 = self.node_vec.len() as u32;
-        self.node_vec.push(NodePtr::Pair(first, rest));
+        let r = self.pair_vec.len() as i32;
+        self.pair_vec.push(IntPair { first, rest });
         r
     }
 
     fn atom<'a>(&'a self, node: &'a Self::Ptr) -> &'a [u8] {
-        match self.node_vec[*node as usize] {
-            NodePtr::Atom(IntAtomBuf { start, end }) => &self.u8_vec[start as usize..end as usize],
-            _ => panic!("expected atom, got pair"),
+        if *node >= 0 {
+            panic!("expected atom, got pair");
         }
+        let atom = self.atom_vec[(-*node - 1) as usize];
+        &self.u8_vec[atom.start as usize..atom.end as usize]
     }
 
     fn buf<'a>(&'a self, node: &'a Self::AtomBuf) -> &'a [u8] {
         &self.u8_vec[node.start as usize..node.end as usize]
     }
 
     fn sexp(&self, node: &Self::Ptr) -> SExp<Self::Ptr, Self::AtomBuf> {
-        match self.node_vec[*node as usize] {
-            NodePtr::Atom(atombuf) => SExp::Atom(atombuf),
-            NodePtr::Pair(left, right) => SExp::Pair(left, right),
+        if *node >= 0 {
```

Comment:
I'm worried about atom/pair type confusion here (clvm types, not rust types).
In the last version, null was type `Atom`, but now, index `0` means "first item in the pair list", if I read this correctly. (And index `-1` is the new null)

> if *node >= 0
Then we return SExp::Pair, but I *think* that was `SExp::Atom` before.

Reply:
Wait, doh. So we still return `SExp::Atom` now.

Reply:
the value `IntAllocator::Ptr` has no relation to the values of anything in clvm. In fact, this type can be an actual pointer too (when using the `ArcAllocator`). The `i32` value of `Ptr` is never available to the clvm program

Reply:
Thanks, that helps!

But what I meant was: when we construct a `SExp` here, if `node >= 0`, we construct a `SExp::Pair`. However, above, we pre-declare `null` on the *atom* stack:

> r.atom_vec.push(IntAtomBuf { start: 0, end: 0 });

Reply:
And then I realized, oh, but the input to this function would be `-1` to mean `null`

Reply:
right